### PR TITLE
Allow Deceased Primary and Opp Acct Change w/ same Default and HH RT

### LINKED ISSUE #2060: Null Pointer Exception if only member of household is marked Deceased and opportunity is added for that contact
Thread: https://powerofus.force.com/0D58000002eBDjh

### LINKED ISSUE #2150: Primary Opportunity Role Deleted when Default Role matches Household Role and HH changes
On the Contact Role Settings Page, if the Opportunity Contact Role Default Role is the same as the Household Member Contact Role and the Opportunity is moved to a new Household, the trigger to create the Household Member Contact Roles is deleting the Primary Contact before adding in the members of the Current Household.

### PR DESCRIPTION
**Pull request by jjbennett**
# Warning
# Info
# Issues
Fixes #2060 
Fixes #2150

### CODE REVIEW

**Comment by jjbennett** [comment]
@njjc when you have a chance, can you please review. Thanks!

File: src/classes/HH_OppContactRoles_TDTM.cls [hunk 1]
```diff
@@ -258,7 +258,7 @@ public class HH_OppContactRoles_TDTM extends TDTM_Runnable {
                                     thisRole.isPrimary = true;
                                     contactRolesForCreation.add(thisRole);
                                 }
-                            } else {
+                            } else if (!householdMember.Deceased__c) {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TDTM.cls]: Nesting too many `IF` statements affects readability and makes maintenance more difficult. Consider refactoring to simplify the flow of this code block.

File: src/classes/HH_OppContactRoles_TDTM.cls [hunk 2]
```diff
@@ -258,7 +258,7 @@ public class HH_OppContactRoles_TDTM extends TDTM_Runnable {
...
                                 // exclude any Contacts with an excluded record type
                                 if (recordTypesOnCons && conRecordTypesToExclude.contains((id)householdMember.get('RecordTypeId'))) {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TDTM.cls]: Nesting too many `IF` statements affects readability and makes maintenance more difficult. Consider refactoring to simplify the flow of this code block.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 3]
```diff
@@ -504,5 +504,143 @@ private class HH_OppContactRoles_TEST {
...
+    */ 
+    static testmethod void testOppsAcctChangeHHToHH() {
+        if (strTestOnly != '*' && strTestOnly != 'testOppsAcctChangeHHToHH') return;
+        testOppsAcctChange(CAO_Constants.HH_ACCOUNT_TYPE, CAO_Constants.HH_ACCOUNT_TYPE);
+    }
+
+    /*******************************************************************************************************
+    * @description Move an Opportunity to a different One2One Account and verifies primary OCR remains
+    */ 
+    static testmethod void testOppsAcctChangeOne2OneToOne2One() {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 4]
```diff
@@ -504,5 +504,143 @@ private class HH_OppContactRoles_TEST {
...
+    */ 
+    static testmethod void testOppsAcctChangeOrgToOne2One() {
+        if (strTestOnly != '*' && strTestOnly != 'testOppsAcctChangeOrgToOne2One') return;
+        testOppsAcctChange(null, CAO_Constants.ONE_TO_ONE_ORGANIZATION_TYPE);
+    }
+
+    /*******************************************************************************************************
+    * @description Move an Opportunity frome One2One to Org Account and verifies primary OCR remains
+    */ 
+    static testmethod void testOppsAcctChangeOne2OneToOrg() {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 5]
```diff
@@ -504,5 +504,143 @@ private class HH_OppContactRoles_TEST {
...
+    */ 
+    static testmethod void testOppsAcctChangeOrgToHH() {
+        if (strTestOnly != '*' && strTestOnly != 'testOppsAcctChangeOrgToHH') return;
+        testOppsAcctChange(null, CAO_Constants.HH_ACCOUNT_TYPE);
+    }
+
+    /*******************************************************************************************************
+    * @description Move an Opportunity from HH to Org Account and verifies primary OCR remains
+    */ 
+    static testmethod void testOppsAcctChangeHHToOrg() {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: * I suggest to annotate test methods with `@IsTest` for better readability.
* Methods should use the `lowerCamelCase` naming convention. Please rename accordingly.

Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: * Methods should use the `lowerCamelCase` naming convention. Please rename accordingly.
* I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 6]
```diff
@@ -504,5 +504,143 @@ private class HH_OppContactRoles_TEST {
             system.assertEquals(0, listSoft[4]);
         }
     }
-       
+
+    /*********************************************************************************************************
+    * @description Tests that an Opportunity Contact Role is created when the deceased member is set as the primary
+    * @return void
+    */
+    private static testMethod void opportunityDeceasedPrimaryAccount() {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls (hunk 3 above)
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls (hunk 4 above)
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls (hunk 6 above)
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 7]
```diff
@@ -504,5 +504,143 @@ private class HH_OppContactRoles_TEST {
...
+                npo02__Household_Member_Contact_Role__c='Household Member'));
+
+        UTIL_UnitTestData_TEST.createAccountContactTestData(CAO_Constants.HH_ACCOUNT_TYPE, 1, 1, 1);
+        Account acct = UTIL_UnitTestData_TEST.listAccT[0];
+        Contact con = UTIL_UnitTestData_TEST.listConT[0];        
+        con.Deceased__c = true;  
+        update con;
+                
+        //create opportunity
+        Opportunity o = new Opportunity(
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: This variable doesn't seem to have a very insightful name. Please give variables names that suggest their purpose.

Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: This variable doesn't seem to have a very insightful name. Please give variables names that suggest their purpose.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 8]
```diff
@@ -504,5 +504,143 @@ private class HH_OppContactRoles_TEST {
...
+    */ 
+    static testmethod void testOppsAcctChangeHHToOrg() {
+        if (strTestOnly != '*' && strTestOnly != 'testOppsAcctChangeHHToOrg') return;
+        testOppsAcctChange(CAO_Constants.HH_ACCOUNT_TYPE, null);
+    }
+
+    /*******************************************************************************************************
+    * @description Move an Opportunity from Org to One2One Account and verifies primary OCR remains
+    */ 
+    static testmethod void testOppsAcctChangeOrgToOne2One() {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 9]
```diff
@@ -504,5 +504,143 @@ private class HH_OppContactRoles_TEST {
...
+        List<OpportunityContactRole> createdRoles = [select Role, ContactId, IsPrimary from OpportunityContactRole where OpportunityId = :o.id order by ContactId];
+        System.assertEquals(1, createdRoles.size(), 'The deceased contact should be added as the primary Contact Role for the Opportunity.');
+        System.assertEquals(contactSettingsForTests.npe01__Opportunity_Contact_Role_Default_role__c, createdRoles[0].Role, 'The Role should match the default contact role.');
+        System.assert(createdRoles[0].IsPrimary, 'The Contact Role should be flagged as primary.');
+    }
+
+    /*******************************************************************************************************
+    * @description Move an Opportunity to a different HH Account and verifies primary OCR remains
+    */ 
+    static testmethod void testOppsAcctChangeHHToHH() {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: * I suggest to annotate test methods with `@IsTest` for better readability.
* Methods should use the `lowerCamelCase` naming convention. Please rename accordingly.

Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: * I suggest to annotate test methods with `@IsTest` for better readability.
* Methods should use the `lowerCamelCase` naming convention. Please rename accordingly.

File: src/classes/HH_OppContactRoles_TEST.cls (hunk 8 above)
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 10]
```diff
@@ -504,5 +504,143 @@ private class HH_OppContactRoles_TEST {
...
+    */ 
+    static testmethod void testOppsAcctChangeOne2OneToOne2One() {
+        if (strTestOnly != '*' && strTestOnly != 'testOppsAcctChangeOne2OneToOne2One') return;
+        testOppsAcctChange(CAO_Constants.ONE_TO_ONE_ORGANIZATION_TYPE, CAO_Constants.ONE_TO_ONE_ORGANIZATION_TYPE);
+    }
+
+    /*******************************************************************************************************
+    * @description MMove an Opportunity from Org to HH Account and verifies primary OCR remains
+    */ 
+    static testmethod void testOppsAcctChangeOrgToHH() {
```
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: * I suggest to annotate test methods with `@IsTest` for better readability.
* Methods should use the `lowerCamelCase` naming convention. Please rename accordingly.

Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: * Methods should use the `lowerCamelCase` naming convention. Please rename accordingly.
* I suggest to annotate test methods with `@IsTest` for better readability.

File: src/classes/HH_OppContactRoles_TEST.cls (hunk 9 above)
Comment (LurchTheButler) [inline src/classes/HH_OppContactRoles_TEST.cls]: * Methods should use the `lowerCamelCase` naming convention. Please rename accordingly.
* I suggest to annotate test methods with `@IsTest` for better readability.

**Pull request merged**